// @mgilangjanuar/telegram/tl/custom/messageButton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageButton = void 0;
const api_1 = require("../api");
const button_1 = require("./button");
const util_1 = require("util");
const Helpers_1 = require("../../Helpers");
class MessageButton {
    constructor(client, original, chat, bot, msgId) {
        this.button = original;
        this._bot = bot;
        this._chat = chat;
        this._msgId = msgId;
        this._client = client;
    }
    [util_1.inspect.custom]() {
        return (0, Helpers_1.betterConsoleLog)(this);
    }
    get client() {
        return this._client;
    }
    get text() {
        return !(this.button instanceof button_1.Button) ? this.button.text : "";
    }
    get data() {
        if (this.button instanceof api_1.Api.KeyboardButtonCallback) {
            return this.button.data;
        }
    }
    get inlineQuery() {
        if (this.button instanceof api_1.Api.KeyboardButtonSwitchInline) {
            return this.button.query;
        }
    }
    get url() {
        if (this.button instanceof api_1.Api.KeyboardButtonUrl) {
            return this.button.url;
        }
    }
    async click({ sharePhone = false, shareGeo = [0, 0] }) {
        if (this.button instanceof api_1.Api.KeyboardButton) {
            return this._client.sendMessage(this._chat, {
                message: this.button.text,
                parseMode: undefined,
            });
        }
        else if (this.button instanceof api_1.Api.KeyboardButtonCallback) {
            const request = new api_1.Api.messages.GetBotCallbackAnswer({
                peer: this._chat,
                msgId: this._msgId,
                data: this.button.data,
            });
            try {
                return await this._client.invoke(request);
            }
            catch (e) {
                if (e.errorMessage == "BOT_RESPONSE_TIMEOUT") {
                    return null;
                }
                throw e;
            }
        }
        else if (this.button instanceof api_1.Api.KeyboardButtonSwitchInline) {
            return this._client.invoke(new api_1.Api.messages.StartBot({
                bot: this._bot,
                peer: this._chat,
                startParam: this.button.query,
            }));
        }
        else if (this.button instanceof api_1.Api.KeyboardButtonUrl) {
            return this.button.url;
        }
        else if (this.button instanceof api_1.Api.KeyboardButtonGame) {
            const request = new api_1.Api.messages.GetBotCallbackAnswer({
                peer: this._chat,
                msgId: this._msgId,
                game: true,
            });
            try {
                return await this._client.invoke(request);
            }
            catch (e) {
                if (e.errorMessage == "BOT_RESPONSE_TIMEOUT") {
                    return null;
                }
                throw e;
            }
        }
        else if (this.button instanceof api_1.Api.KeyboardButtonRequestPhone) {
            if (!sharePhone) {
                throw new Error("cannot click on phone buttons unless sharePhone=true");
            }
            const me = (await this._client.getMe());
            const phoneMedia = new api_1.Api.InputMediaContact({
                phoneNumber: me.phone || "",
                firstName: me.firstName || "",
                lastName: me.lastName || "",
                vcard: "",
            });
            throw new Error("Not supported for now");
            // TODO
            //return this._client.sendFile(this._chat, phoneMedia);
        }
        else if (this.button instanceof api_1.Api.InputWebFileGeoPointLocation) {
            if (!shareGeo) {
                throw new Error("cannot click on geo buttons unless shareGeo=[longitude, latitude]");
            }
            let geoMedia = new api_1.Api.InputMediaGeoPoint({
                geoPoint: new api_1.Api.InputGeoPoint({
                    lat: shareGeo[0],
                    long: shareGeo[1],
                }),
            });
            throw new Error("Not supported for now");
            // TODO
            //return this._client.sendFile(this._chat, geoMedia);
        }
    }
}
exports.MessageButton = MessageButton;
